File: src/main/scala/com/flink/tutorials/api/transformations/MapExample.scala
package com.flink.tutorials.api.transformations

import org.apache.flink.api.common.functions.RichMapFunction
import org.apache.flink.streaming.api.scala._

object MapExample {
  def main(args: Array[String]): Unit = {
    // 创建 Flink 执行环境
    val senv: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment

    val dataStream: DataStream[Int] = senv.fromElements(1, 2, -3, 0, 5, -9, 8)

    // Lambda函数 =>
    val lambda = dataStream.map ( input => input * 2 ).print()

    // Lambda函数 _
    val lambda2 = dataStream.map { _ * 2}.print()

    // 继承RichMapFunction
    // 第一个参数是输入，第二个参数是输出
    class DoubleMapFunction extends RichMapFunction[Int, Int] {
      def map(in: Int):Int = { in * 2 }
    };

    // 匿名函数
    val anonymousFunction = dataStream.map {new RichMapFunction[Int, Int] {
      def map(input: Int): Int = {
        input * 2
      }
    }}.print()

    val richFunction = dataStream.map {new DoubleMapFunction()}.print()

    senv.execute("Basic Map Transformation")
  }
}

File: src/main/java/com/flink/tutorials/java/chapter4/transformations/MapExample.java
package com.flink.tutorials.java.chapter4.transformations;

import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;


public class MapExample {

    public static void main(String[] args) throws Exception {

        // 创建 Flink 执行环境
        StreamExecutionEnvironment senv = StreamExecutionEnvironment.getExecutionEnvironment();


        DataStream<Integer> dataStream = senv.fromElements(1, 2, -3, 0, 5, -9, 8);

        // 使用Lambda表达式
        DataStream<String> lambdaStream = dataStream
                .map(input -> "lambda input : " + input + ", output : " + (input * 2));
        lambdaStream.print();

        DataStream<String> functionDataStream = dataStream.map(new DoubleMapFunction());
        functionDataStream.print();

        // 匿名类
        DataStream<String> anonymousDataStream = dataStream.map(new MapFunction<Integer, String>() {
            @Override
            public String map(Integer input) throws Exception {
                return "anonymous function input : " + input + ", output : " + (input * 2);
            }
        });
        anonymousDataStream.print();

        senv.execute("basic map transformation");
    }

    // 继承MapFunction
    // 第一个泛型Integer是输入类型，第二个泛型String是输出类型
    public static class DoubleMapFunction implements MapFunction<Integer, String> {
        @Override
        public String map(Integer input) {
            return "function input : " + input + ", output : " + (input * 2);
        }
    }
}
